### ISSUE TITLE
Files: Persistent Identifiers for Files (DOIs for files)

### ISSUE THREAD

**Issue by eaquigley** [IQSS/dataverse#2438]
Since we are moving towards individual pages for files, we also need to think about what the persistent identifier will be for them.

**Comment by lwo** [comment]
I am entering via: https://groups.google.com/forum/#!msg/dataverse-community/gtz2npccWjU/i7_EVs2LBgAJ
 
... I think persistent identifiers should not be derived from the local file id. An organization may want to migrate from their repository solution to dataverse and be able to import their PIDs if they already have them. And then rebind them to the dataverse dataset and future file pages. That would be possible with a String type, but not with a number.

**Comment by djbrooke** [comment]
Sebastian brought this up on the community call 8/16 and asked if it would be included in 4.6. There are currently no plans to work on this for 4.6.

**Comment by mheppler** [comment]
Closing this issue as a duplicate of Files: Need Persistent Identifiers/URL's for Data Files #2700.

**Comment by pdurbin** [comment]
A "DOIs for files" thread was just started at https://groups.google.com/d/msg/dataverse-community/JX2GLqPy_yE/6dzgXGVcCAAJ which reminds me that @djbrooke and I discussed this issue as well as #2700 last week.

In short, #2700 was a bit more about a specific need for putting into print specific instructions about how to download files from an installation of Dataverse. A combination of a DOI and a file name was sufficient so I'll close that issue.

This issue was originally closed as a duplicate of #2700 but this issue is actually the "DOIs for files" issue so I'll reopen it. Note that the title is "persistent identifiers for files" to be more generic than DOIs, to include Handles or whatever other schemes make sense.

**Comment by scolapasta** [comment]
Two recent discussions come up recently where we would benefit from persistent IDs for files:

-supporting the Data Citation recommendations for granularity
- using the persistent identifiers for provenance (as provenance should never change)

**Comment by djbrooke** [comment]
Thanks @mheppler for offering to drop in a quick mockup here.

There are some open questions about the ordering of the items in the file citation, what's included, and how to include both the file DOI and dataset DOI. @mercecrosas and @scolapasta will connect about this.

**Comment by scolapasta** [comment]
Discussed with @mercecrosas some of the open questions - she is going to RDA next week and will be able to consult with Martin Fenner from Datacite, so answers may have to wait until then, but we should be able to get started on some of the logic for this in the meanwhile.

**Comment by mheppler** [comment]
Here is the mockup of the file pg with the updated file citation format, and new File Persistent ID displayed.

![file-doi-filepg](https://user-images.githubusercontent.com/687227/30560716-bf0da900-9c86-11e7-9c9e-e224b09870a3.png)

**Comment by mheppler** [comment]
**UI IMPACT**

- [ ] Citation block styling: gray outline+background, added small file icon (also add small dataset icon to citation block on dataset pg)
- [ ] Citation format: moved file name up, before the DOI, moved dataset name to the end, as a suffix with a new "datasetName" bracketed label.
- [ ] File metadata: added new "File Persistent ID" label and value

**Comment by scolapasta** [comment]
After further consulting with @mercecrosas a few additional points:

1. There will be one DOI (or other persistent identifier) for each file in dataverse. As an example, if you replace a file, that new file would have a separate DOI.
2. We're still determining what we will use for the identifier, but it will likely be based on the DOI for the dataset
3. We will not change the citation as above. Rather we will keep the citation for files as is, but aldd the doi at the end similar to the name with [fileDOI].

For example:

LastName, FirstName, 2017, "Dataset", doi:10.5072/FK2/XXXXXX, Demo Dataverse, V1; filename.txt [fileName] doi:10.5072/FK2/XXXXXX/YYY

**Comment by adam3smith** [comment]
I'm excited you're working on this, thank you!
As requested by Gustavo, a couple of comments.

About the citation: I do think having two DOIs is unwieldly -- e.g. most reference managers are not going to be able to produce such citations, and they're going to be quite long. (As an aside, the DOI display guidelines ([CrossRef](https://www.crossref.org/display-guidelines/), [Datacite](https://support.datacite.org/v1.0/docs/datacite-doi-display-guidelines) now universally discourage the use of doi: and favor https://doi.org/. I think Dataverse should switch its default. It's one of the things we customize in our install)

While the suggested citation is more a question of taste, this:

> We're still determining what we will use for the identifier, but it will likely be based on the DOI for the dataset

is bad practice. DOIs should not contain human-readable ("semantic") elements; the connection between file & data is handled in the metadata (`isPartOf` in the Datacite kernel), not in the DOI itself, and the DOI should not suggest to users that it is. See e.g. Martin Fenner on the topic [here](https://blog.datacite.org/cool-dois/#suffix).

**Comment by jggautier** [comment]
Working on a draft [DataCite metadata template for files](https://docs.google.com/document/d/1tevjRIX96EFCYzAgw7KU-bmKT_kabyZb-ilMHxyb5dc/edit).

**Comment by rbhatta99** [comment]
latest Update before me handing this over:
The latest commit is broken for now, as a lot of changes are needed to made to properly generate and register DOIs for individual datafiles. There are placeholders in various DOI/handle registration beans, which will take in DvObjects instead of datasets, which will accomodate for the registration of the datafiles. The current state of the branch is:

- A new command has been written for the creation and saving of DataFiles (namely CreateDataFileCommand), which works fine as long as the files are not sent for registration and DOI generation
- @jggautier has provided a template for the metadat of a DataFile, which needs to be sent over to Datacite for registration.
- I have written a couple of functions in abstractIdServiceBean which takes in DvObjects instead of datasets, but have commented them out, as call a lot of functions, which in themselves use Datasets as arguments, and need to be changed to accomodate DvObjects.

**Comment by mercecrosas** [comment]
Sebastian,

Apologies it took me so long to reply. You bring up very useful comments:

- On the File citation format with 2 DOIs: I agree with you are pointing
out. We'll revisit it based on your comments. Since the beginning of
Dataverse, we have tried to follow the format proposed by Altman & King
2007 (https://gking.harvard.edu/files/abs/cite-abs.shtml), but at the same
times we want to be compliant with the Data Citation implementation roadmap
as well as, as you say, follow a standard that works well for reference
managers.
*So, based on all this, do you have suggestions for the format for the
file-level citation that Dataverse can use? *(of course, the citation
metadata will include the file and the dataset DOI with the proper
relationship, but we are here focusing on the format we display for
file-level citation).

- DO should not contain semantic elements: I agree and disagree. I know
that Martin is a big advocate of this (ID should be opaque, and not contain
human-readable-elements), and I understand well why is a good idea.
However, user-friendly IDs do help users to quickly have an idea of what
that object means, how it relates to others, etc. The system, of course,
should not rely on this and even users should understand that what is
important is the metadata and not to use any semantics in the ID as a
guarantee of what it is. But still, I think it's useful for users. Dryad
uses something similar, and at a first glance it seems useful. That's why
we propose it. If the community feel strongly that we should use arbitrary
IDs, we can change this. *I'm interested in hearing opinions from the
community!*

Merce


----------
Mercè Crosas, Ph.D., Chief Data Science and Technology Officer, IQSS, Harvard
University
T:@mercecrosas W:mercecrosas.com

On Thu, Sep 28, 2017 at 2:32 PM, Sebastian Karcher <notifications@github.com

**Comment by pdurbin** [comment]
I got a quick brain dump from @sekmiller and plan to code review his pull request: #4224

**Comment by TaniaSchlatter** [comment]
Here's a revised mockup:
<img width="1023" alt="img_6156" src="https://user-images.githubusercontent.com/15386424/31970404-d12498ee-b8e5-11e7-8d73-59bb5d3ece15.PNG">

**Comment by jggautier** [comment]
Just some more clarifying notes after talking with @TaniaSchlatter: 

- The line below the file name should read "File PID" or ("Persistent ID"?), and of course followed by the file PID. 
- The file page URL there now is just a placeholder.
- If the PID is a DOI (in some installations it won't be), it should be a URL, e.g. https://doi.org/10:7910/DVN/...

**Comment by mheppler** [comment]
Discussed the mockup @TaniaSchlatter added compared to the UI work that I had already to this branch and found a common ground that looks something like the screenshot attached. Sending to Code Review.

![screen shot 2017-10-25 at 5 25 41 pm](https://user-images.githubusercontent.com/687227/32024294-f4932c5e-b9a9-11e7-84c6-5ad59504aa59.png)

**Comment by landreev** [comment]
What is the plan for the existing production DataFiles - they will need to be assigned individual identifiers, correct? Will there be a migration script? Is it going to be handled as a different issue?

**Comment by landreev** [comment]
I've moved the ticket into QA. I've done as much reviewing of it as can be done without doing a full QA on my own, I think. My main remaining question: the code that handles finalizing adding the newly created datafiles to the dataset, and saving them on the storage permanently, have been moved from the ingest service and into the create datafile command. So it is now being done for every file individually, as opposed to the way it was being done, on a list of files, at all once. From just looking at the code, it's not immediately clear to me whether there is any performance hit related to this. So I recommend that we specifically test this during QA - try adding 1000 files via drag and drop, and see if there is any difference in performance, before vs after.

**Comment by djbrooke** [comment]
@landreev good catch(es). Yes we should provide these DOIs for existing files. There should be necessary migration scripts provided with this issue and we should provide it for any type of PIDs used by the community. 

@landreev @kcondon @scolapasta do you think we should move this to QA and work on the migration scripts in parallel or should it go back to dev?

**Comment by landreev** [comment]
@kcondon @scolapasta @djbrooke 
There's enough working functionality in this issue as is, that can be QA-ed. So I'm not sure it would be justified, to delay it until the migration process is added. 

That said, I did not work on this issue myself, and I will not be doing the QA - so mine is definitely not the authoritative opinion on this...

**Comment by landreev** [comment]
@kcondon @scolapasta @djbrooke
(rather than moving the issue back into dev., opening a new issue for the migration process may be more in line with the "small chunks" approach - ?)

**Comment by djbrooke** [comment]
If we need the migration script for this to be released and to have benefit to the community, let's move it back to dev and add that script.

**Comment by pdurbin** [comment]
I hate to say it, but now that 4.8.3 is out `upgrade_v4.8.2_to_v4.8.3.sql` should be renamed yet again.

**Comment by pdurbin** [comment]
From standup, it sounds like this issue is still being worked on. We are trying to incorporate feedback from partners into the `2438-persistent-identifiers-for-files` branch.

I did go ahead and fix the SQL script filename in 3166074

**Comment by sekmiller** [comment]
For testing the following sql script will create the new fields on the dvObject table, populate them from the values on the dataset table, and remove the fields from the dataset table:
scripts/database/upgrades/upgrade_v4.8.3_to_v4.8.4.sql

To populate identifiers in existing files there is an api:
http://$SERVER/api/admin/registerDataFileAll
which requires a superuser token

When any kind of datafile (except for a file uploaded specifically as a thumbnail file) a DOI should be assigned. If you are setup for EZID you can verify that a DOI has been created for a file by using this link:
https://ezid.cdlib.org/id/doi:10.5072/FK2/MFS2BW (with the file's DOI - which can be viewed on the file landing page.)

If a dataset containing a given file is destroyed, the DOIs for the corresponding files should be destroyed as well. If a file is removed from a subsequent version its DOI should remain. If a file is deleted before being published its DOI should be deleted.

**Comment by pdurbin** [comment]
@sekmiller discussed how it might be handy to test both #2438 and #4295 in a single branch so I created a `2438-4295-dois-for-files` branch. I started with the `2438-persistent-identifiers-for-files` branch and created that new branch. Then I merged `4295-File-download-api-via-persistent-identifiers` into it. Then I merged `develop` into it. @sekmiller said he'd take a look. He can make a pull request and do the "connects to" stuff in Waffle if he thinks this is better.

**Comment by dlmurphy** [comment]
Just committed updated documentation for this issue in f97e87ca34caa527151fd08787427f2057314c79. Added a long overdue user guide section about the File Page, and made sure file DOI related sections were accurate across all guides. Documentation is ready for review!

**Comment by kcondon** [comment]
-Need to add doc for 2 new api endpoints
-File DOIs do not resolve to file landing page, gives 404
-Uploading a large number of files (1000) is considerably slower (11 mins versus 1 min) when registering (handle/ezid on create, datacite on publish).
-Scrollbar on file table does not work with 1000 files, not present. Works in /develop.

**Comment by djbrooke** [comment]
Thanks @kcondon for the numbers on the performance hit. Do we see a similar (percentage) increase on 10 or 50 files?

**Comment by kcondon** [comment]
Good question

**Comment by pdurbin** [comment]
> Uploading a large number of files (1000) is considerably slower (7-11 mins versus 1 min) when registering

@sekmiller asked me to take a look at this.

**Comment by pdurbin** [comment]
One observation I'll make is that as of d75e866 when I click a file in a search card I land on a URL with `persistentId` in the URL (i.e. http://localhost:8080/file.xhtml?persistentId=doi:10.5072/FK2/TSA7BQ ) but when I click the same file on the dataset page I see `fileId` in the URL instead ( i.e. http://localhost:8080/file.xhtml?fileId=237&version=DRAFT&version=. ). Do we want these to be consistent? I assume so. (Also note `version=.` which is very strange, perhaps related to the strange `version=.0` reported at #4308.) I'm logged in as `dataverseAdmin` if that matters.

**Comment by pdurbin** [comment]
Uploading 1000 small text files as of d75e866 (via zip, attached) and then clicking "Save Changes" is definitely slow, taking about 8 minutes, as reported above. In addition, these three lines appear in `server.log` for each of the 1000 files:

```
[2017-12-12T10:45:18.123-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.core.index.DOIEZIdServiceBean] [tid: _ThreadID=106 _ThreadName=http-listener-1(4)] [timeMillis: 1513093518123] [levelValue: 900] [[
  String edu.ucsb.nceas.ezid.EZIDException: bad request - unrecognized identifier scheme]]

[2017-12-12T10:45:18.123-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.core.index.DOIEZIdServiceBean] [tid: _ThreadID=106 _ThreadName=http-listener-1(4)] [timeMillis: 1513093518123] [levelValue: 900] [[
  localized message bad request - unrecognized identifier scheme]]

[2017-12-12T10:45:18.123-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dvn.core.index.DOIEZIdServiceBean] [tid: _ThreadID=106 _ThreadName=http-listener-1(4)] [timeMillis: 1513093518123] [levelValue: 900] [[
  cause]]
```

Finally, the files are not indexed but this is a known issue, first reported at #3243 (I created the dataset first and then later uploaded the files).

Here's the zip file I used to test: [1000uniquefiles.zip](https://github.com/IQSS/dataverse/files/1552266/1000uniquefiles.zip) (created with `for i in {1..1000}; do echo $i > file$i.txt; done`).

**Comment by pdurbin** [comment]
I re-tested the same zip file with 1000 small files above on the `develop` branch as of 3a53923 and it was much faster. "Only" 30 seconds or so instead of ~8 minutes.

Also, I showed this to @kcondon already but I'm seeing this stack trace over and over in `server.log` just after uploading the zip file in the GUI but before clicking "Save Changes":

```
[2017-12-12T11:33:33.775-0500] [glassfish 4.1] [INFO] [] [edu.harvard.iq.dataverse.ingest] [tid: _ThreadID=30 _ThreadName=http-listener-1(1)] [timeMillis: 1513096413775] [levelValue: 800] [[
  cause.getMessage() was null for java.lang.reflect.InvocationTargetException]]

[2017-12-12T11:33:33.775-0500] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=30 _ThreadName=Thread-9] [timeMillis: 1513096413775] [levelValue: 1000] [[
  java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor640.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at edu.harvard.iq.dataverse.ingest.IngestableDataChecker.detectTabularDataFormat(IngestableDataChecker.java:592)
	at edu.harvard.iq.dataverse.util.FileUtil.determineFileType(FileUtil.java:290)
	at edu.harvard.iq.dataverse.util.FileUtil.createDataFiles(FileUtil.java:810)
	at edu.harvard.iq.dataverse.EditDatafilesPage.handleFileUpload(EditDatafilesPage.java:1794)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.el.parser.AstValue.invoke(AstValue.java:289)
	at com.sun.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:304)
	at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
	at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
	at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:105)
	at org.primefaces.component.fileupload.FileUpload.broadcast(FileUpload.java:318)
	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:755)
	at javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:931)
	at com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:78)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:646)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:295)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:205)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:415)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:201)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:175)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:561)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:565)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:545)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.BufferUnderflowException
	at java.nio.ByteBuffer.get(ByteBuffer.java:688)
	at java.nio.DirectByteBuffer.get(DirectByteBuffer.java:285)
	at edu.harvard.iq.dataverse.ingest.IngestableDataChecker.testRDAformat(IngestableDataChecker.java:549)
	... 59 more]]
```

**Comment by pdurbin** [comment]
Somewhat unsurprisingly, the performance problem seems to be in sending the requests over the wire to EZID to reserve a DOI one by one for each of the 1000 files. I'm not sure if there's a quick fix for this:

![screen shot 2017-12-12 at 3 22 31 pm](https://user-images.githubusercontent.com/21006/33906630-7d19b324-df50-11e7-8fdd-033b10947eb2.png)

**Comment by kcondon** [comment]
I wonder what it does for DataCite since that should not be sending requests on create/upload and it exhibits the same slowness?

**Comment by pdurbin** [comment]
@kcondon good thought. I can try to switch to DataCite and repeat the same test. Before I move of of EZID however, let me show another interesting finding. It appears that quite a bit of time is also being spent in `generateDatasetIdentifier` (23%), the `alreadyExists` method. @sekmiller suspects this may be a file-level operation and that we could simply rename the method if it's confusing:

![screen shot 2017-12-12 at 3 45 22 pm](https://user-images.githubusercontent.com/21006/33907770-4af53504-df54-11e7-97e7-6977258dc865.png)

**Comment by pdurbin** [comment]
Ok, please ignore the actual time spent because I took a screenshot before "Save Changes" finished but for DataCite, the time (yes, it's slow) is being spent in the `testDOIExists` method which reaches out over the wire to DataCite using an HTTP client (`org.apache.http.impl.client.CloseableHttpClient`):

![screen shot 2017-12-12 at 4 27 39 pm](https://user-images.githubusercontent.com/21006/33909420-9a49eea6-df59-11e7-9391-203d74f06db6.png)

**Comment by pdurbin** [comment]
@sekmiller and I did a little brainstorming on potential fixes to the performance problem of registering 1000 files with an external service like EZID or DataCite or Handle.

On "Save Change" after upload or on publish (when reserved persistent identifiers are made public or whatever) we don't want to see the spinner in the GUI for 8 minutes:

![screen shot 2017-12-12 at 4 31 34 pm](https://user-images.githubusercontent.com/21006/33909538-fad314e6-df59-11e7-8aa8-e5e38d3455c3.png)

On "Save Changes" it would be nice to lock the dataset and let the GUI return right away. In the background, the DOIs for each of the 1000 files could be reserved. There are a few places in the app where we do background processing like this:

- ingest (JMS)
- grant role (indexing is called with `@Asynchronous` if memory serves)
- the "crawl filesystem" code before we introduce the "data package" concept (Java EE Batch API)
- the SBGrid "lock dataset on publish while me move files from 'hold' to 'public'" concept (Workflows)

Maybe there are others? We need to make a decision on if we're going to try one of the approaches above or another approach in the current sprint or defer this work until later on.

By the way, I forget what the user experience is for SBGrid when you click "Publish" and have to wait for a lot of background processing but I assume we'd want the same user experience. That is, we wouldn't want to say that the dataset is published until each file has a public DOI or Handle.

**Comment by scolapasta** [comment]
I'd nominate spinning off the performance as a separate issue. That doesn't necessarily mean that it is not part of this pull request / merge, just that as a separate issue, we can separate out all comments related to just performance.

(Part of my reasoning is that I don't want my next comment to get lost :)

**Comment by scolapasta** [comment]
Discussed further @mercecrosas and for the identifier we've decided that we will make it configurable per installation. Initially we will provide two different styles, one will be an arbitrary identifiers, as datasets currently used, and the other will be one that will build on the identifier of the dataset (e.g. 1234/567 for a dataset with identifier 1234).

As @mercecrosas says above:
"user-friendly IDs do help users to quickly have an idea of what
that object means, how it relates to others, etc. The system, of course,
should not rely on this and even users should understand that what is
important is the metadata and not to use any semantics in the ID as a
guarantee of what it is. But still, I think it's useful for users. Dryad
uses something similar, and at a first glance it seems useful. "

**Comment by kcondon** [comment]
OK file urls:
From file card:
https://dataverse-internal.iq.harvard.edu/file.xhtml?persistentId=doi:10.5072/Q69UCB
From dataset files tab:
https://dataverse-internal.iq.harvard.edu/file.xhtml?persistentId=doi%3A10.5072%2FQ69UCB&version=2.0

So, they both now use the persistentId form as requested but I'm told the escaped format of the dataset files tab URL is due to how the backend processing works. Note that adding &version=2.0 to the end of the file card url (no escaping needed) also works.

**Comment by pdurbin** [comment]
@kcondon good catch. The same thing happens at the dataset level (and has for a long time) when you create a dataset. In the URL you see `%3A` for a colon (:) and `%2F` for a slash (/). I'm not sure if it's in scope to fix it at the dataset level as well or not.

**Comment by pdurbin** [comment]
The code in question that was generating a URL with `persistentId=doi%3A10.5072%2FFK2%2FP3ENAC` (ugly, percent-encoded) vs `persistentId=doi:10.5072/FK2/P3ENAC` (clean, not-percent encoded) was at https://github.com/IQSS/dataverse/blob/18eb900214815b1c666508c70f714af9dab402f9/src/main/webapp/filesFragment.xhtml#L221 and uses `h:outputLink`, like this:

```
<h:outputLink id="fileNameLink" value="#{widgetWrapper.wrapURL('/file.xhtml')}">
    <f:param name="persistentId" value="#{fileMetadata.dataFile.globalId}"/>
    <f:param name="version" value="#{fileMetadata.datasetVersion.friendlyVersionNumber}"/>
    <h:outputText value="#{fileMetadata.label}" />
</h:outputLink>
```

I can easily reproduce the same behavior with a more minimal `h:outputLink` example with hard-coded values:

```
<h:outputLink id="testOutputText" value="/dataset.xhtml">
    <f:param name="persistentId" value="doi:10.5072/FK2/LNDANK"/>
    <h:outputText value="h:outputLink does percent encoding" />
</h:outputLink>
```

https://stackoverflow.com/questions/24733959/houtputlink-value-escaped seems related.

The solution I implemented in 2ba7d9a was to replace the `h:outputLink` version with a simple `a href` version:

```
<a href="/file.xhtml?persistentId=#{fileMetadata.dataFile.globalId}&amp;version=#{fileMetadata.datasetVersion.friendlyVersionNumber}">
    #{fileMetadata.label}
</a>
```

I did look briefly at similar behavior on "create dataset" but the code looks complicated uses a different JSF/Primefaces tag (`p:commandButton`) and I don't think we should increase the scope of this issue any more.

**Comment by pdurbin** [comment]
@scolapasta noticed that I removed the call to `widgetWrapper.wrapURL` so I just added it back in 76982ce

**Comment by pdurbin** [comment]
Whoops, re-added with query params in the right place at b637c43

**Comment by pdurbin** [comment]
At standup I mentioned that the "Download URL" link on the file landing page still shows a number like this:

![screen shot 2017-12-18 at 2 21 17 pm](https://user-images.githubusercontent.com/21006/34123829-d527fdc0-e3fe-11e7-9be7-62e89f3dbb8f.png)

We agreed that this should be a DOI instead but I poked around at the code a bit and the change is more extensive than I expected. @sekmiller and I agreed that I'd push a commit to a new branch so he could get a sense of what the change would look like. Here it is: 4f62855

**Comment by kcondon** [comment]
Functional changes/ updates to be tested:
1. APIs are updated to support file doi:
Basic access URI:
/api/access/datafile/$id
All formats tab data bundle download:
/api/access/datafile/bundle/$id
Metadata/ddi access:
/api/access/datafile/$id/metadata/ddi
Restrict a file:
PUT http://$SERVER/api/files/{id}/restrict
Replace a file:
POST http://$SERVER/api/files/{id}/replace?key=$apiKey
Out of scope:
download multiple files, anything on the /meta path, sword delete file, sword get list of files.

2. Configurable change in form of file doi:
INSERT INTO setting(
           name, content)
   VALUES (':DataFilePIDFormat', 'DEPENDENT');

   INSERT INTO setting(
           name, content)
   VALUES (':IdentifierGenerationStyle', 'sequentialNumber');

They work together: DEPENDENT, or INDEPENTDENT determines whether the file doi uses <dataset doi>/identifier or just generates a standalone doi as it does now. sequentialNumber was added to support SBGrid but default is random, need to find the keyword for that. All of this should now be doc'd.

**Comment by sekmiller** [comment]
For the creation of PIDs of DataFiles we added a new setting ":DataFilePIDFormat", the default value is DEPENDENT. As such the PID of any DataFile will pre-pend its Dataset's identifier onto the file identifier. That is, if the Dataset doi is doi:10.5072/FK2/ABC123 the datafiles in that dataset will have DOIs that look like this: doi:10.5072/FK2/ABC123/XYZ789, doi:10.5072/FK2/ABC123/QRS456, etc.

In order to not pre-pend the dataset identifier the setting should be entered as INDEPENDENT. If that is the case the dataset's DOIs will not include the dataset first, so in the example above the file DOIs would be:  doi:10.5072/FK2/XYZ789, doi:10.5072/FK2/QRS456.

**Comment by sekmiller** [comment]
@kcondon, the setting for random/sequential is ":IdentifierGenerationStyle"

**Comment by kcondon** [comment]
I mean the value for :IdentifierGenerationStyle where it is not sequential:
randomString

**Comment by kcondon** [comment]
Found an issue adding files using Dependent/Sequential. Throws exception:
A system exception occurred during an invocation on EJB DatasetServiceBean, method: public java.lang.Long edu.harvard.iq.dataverse.DatasetServiceBean.getMaximumExistingDatafileIdentifier(edu.harvard.iq.dataverse.Dataset)

**Comment by pdurbin** [comment]
@kcondon good catch. Fixed in 15771f3

**Comment by kcondon** [comment]
The dependent/sequential issue is partly fixed: it is fixed if the dataset is created first, then add files. If throws a different error if you attempt to create a dataset and add files at the same time:

[2017-12-21T13:23:53.639-0500] [glassfish 4.1] [SEVERE] [] [edu.harvard.iq.dataverse.DatasetPage] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1513880633639] [levelValue: 1000] [[
  CommandException, when attempting to update the dataset: Can't execute command edu.harvard.iq.dataverse.engine.command.impl.CreateDataFileCommand@5587b40d, because request [DataverseRequest user:[AuthenticatedUser identifier:@kc122017]@140.247.116.74] is missing permissions [EditDataset] on Object depseq]]

[2017-12-21T13:23:53.668-0500] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.DatasetPage] [tid: _ThreadID=51 _ThreadName=jk-connector(2)] [timeMillis: 1513880633668] [levelValue: 900] [[
  No such dataset: doi:10.5072/FK2/8]]

**Comment by pdurbin** [comment]
As of 10dab22 (prior to the the fix I mentioned above, 15771f3) after running `curl http://localhost:8080/api/admin/settings/:IdentifierGenerationStyle -X PUT -d sequentialNumber` I'm getting an error when creating a dataset via GUI (works via API):

![screen shot 2017-12-21 at 1 27 11 pm](https://user-images.githubusercontent.com/21006/34273300-c1719880-e661-11e7-8251-7ea62fb62b4e.png)

`[2017-12-21T13:26:58.186-0500] [glassfish 4.1] [SEVERE] [] [edu.harvard.iq.dataverse.DatasetPage] [tid: _ThreadID=142 _ThreadName=http-listener-1(3)] [timeMillis: 1513880818186] [levelValue: 1000] [[
  CommandException, when attempting to update the dataset: Dataset with identifier ‘4’, protocol ‘doi’ and authority ‘10.5072/FK2’ already exists]]`

It seems to be coming from this part of CreateDatasetCommand

        if ((importType != ImportType.MIGRATION && importType != ImportType.HARVEST) && !ctxt.datasets().isIdentifierUniqueInDatabase(theDataset.getIdentifier(), theDataset, idServiceBean)) {
            throw new IllegalCommandException(String.format("Dataset with identifier '%s', protocol '%s' and authority '%s' already exists",
                    theDataset.getIdentifier(), theDataset.getProtocol(), theDataset.getAuthority()),
                    this);
        }

I'm not sure what's going on. I may drop my database and see if the error goes away.

**Comment by pdurbin** [comment]
I can't reproduce the error after dropping my database but I took a dump of it in case it's helpful: [dataverse_db.sql.txt](https://github.com/IQSS/dataverse/files/1580513/dataverse_db.sql.txt)

**Comment by pdurbin** [comment]
While I was looking around in `DatasetServiceBean.java` I noticed that there's a query that is operating against the dataset table but I think it's supposed to be using the dvobject table instead since columns like `identifier`, `protocol`, and `authority` were moved there:

    public boolean isIdentifierUniqueInDatabase(String userIdentifier, Dataset dataset, IdServiceBean idServiceBean) {
        String query = "SELECT d FROM Dataset d WHERE d.identifier = '" + userIdentifier + "'";
        query += " and d.protocol ='" + dataset.getProtocol() + "'";
        query += " and d.authority = '" + dataset.getAuthority() + "'";

I'm not sure if this has anything to do with the bugs above, but I thought I'd at least mention it.

**Comment by pdurbin** [comment]
Actually, somewhere else in the code this works fine but I'm not sure why since those columns were moved to the `dvobject` table: `String queryStr = "SELECT s from Dataset s where s.identifier = :identifier  and s.protocol= :protocol and s.authority= :authority";` . Dataset extends DvObject so I guess JPA is able to find the columns or something.

**Comment by sekmiller** [comment]
Fixed this issue which was caused by the CreateDataFileCommand having a required permission of EditDataset. Now it will require addDataset or EditDataset depending on whether files are added on create or edit.

Issues with new id structure:

Adding files on dataset create results in 500 error, happens in all combinations of configs:
[2017-12-21T13:23:53.639-0500] [glassfish 4.1] [SEVERE] [] [edu.harvard.iq.dataverse.DatasetPage] [tid: _ThreadID=52 _ThreadName=jk-connector(3)] [timeMillis: 1513880633639] [levelValue: 1000] [[
CommandException, when attempting to update the dataset: Can't execute command edu.harvard.iq.dataverse.engine.command.impl.CreateDataFileCommand@5587b40d, because request [DataverseRequest user:[AuthenticatedUser identifier:@kc122017]@140.247.116.74] is missing permissions [EditDataset] on Object depseq]]

**Comment by michbarsinai** [comment]
All in all looks good. Back to @ferrys for some clean-ups based on GitHub code comments.

**Comment by kcondon** [comment]
Need to combine the old v4.8.4 to v4.9 script into the new v4.8.5 to v4.9 script.

**Comment by sekmiller** [comment]
I updated the migration script as requested. Also, I think that the documentation for the adding DOIs to existing files belongs in the release notes. It's not something that anyone would run outside of a deployment.

**Comment by djbrooke** [comment]
Had a discussion after standup about the performance hit. We should resolve the performance as part of this. We'll do a technical investigation first. @scolapasta is going to check in with some PID-knowledgeable folks.

**Comment by pameyer** [comment]
I did a quick run through with package files (fresh VM).  create API works file; DCM success works file; package file is visible on the dataset page under the files tab. 

Hovering over the URL shows a url of `files.xhtml?persistentId=null:null/null&version=DRAFT` ; unsurprisingly clicking this link leads to an error.  Checking `dvobject` and `datafile` tables shows that DB doesn't have an identifier for the package file.

Did not proceed with run through (submit for review, return from review, publish w\ RSAL, etc).

**Comment by djbrooke** [comment]
Thanks for the discussion after standup about this. 

- Is the cause of the performance issues the communication with the DOI providers? Or is it due to changes in the code that we made? @landreev will investigate this. 
- We do want PIDs for package files, even though it's somewhat redundant right now. In the future, package files will be able to able to live alongside HTTP-uploaded files. This is an easy change.

**Comment by mercecrosas** [comment]
can't wait for @landreev investigationsresults!

**Comment by pameyer** [comment]
As of 915c9fa1ff2208e3b0f9ab585ec37bb110651566 , package files get PIDs - thanks @sekmiller

**Comment by djbrooke** [comment]
Had a quick discussion after standup today:

- We should index the File PIDs so they can be found in search
- We should include File PIDs in Import and Export

**Comment by landreev** [comment]
To reiterate, I'm only investigating one specific area - whether the performance degradation we are seeing is entirely/mostly due to having to make those registration calls to the naming authority. We have reasons to assume that this is the case; but I wanted to make sure. (We used to do certain things - moving files around, etc., in single batch passes; now these tasks have been moved into the CreateFileCommand, which is called separately for each file... so there is at least some potential for the performance hit). 
Will report shortly.

**Comment by landreev** [comment]
One idea, as a followup to the discussion we've just had, about the global ids vs. export: 
- When we add the global ids to the file sections of the JSON and DDI exports, let's add them in the form of the resolvable global ID url. (e.g. "http://dx.doi.org/10.5072/FK2/XXX", and  NOT "doi:10.5072/FK2/")
- By doing this, we can achieve several things:
- We'll have this global URL shown *somewhere*. (according to @kcondon, it's not shown anywhere on any pages)
- We'll solve the issue a partner (Solomon?) recently complained about: since the DDIs presently show the Access API url, by following those, users can download public files without ever seeing any Terms of Use that may be associated with the files. (remember - we are still not enforcing the terms of use on API downloads - it is done on the page level only!). If we replace these URLs with the global id URLs, those resolve to the file.xhtml pages - and that should (somewhat) solve the problem. Knowing the global id, one can still cook the access API url... but it's not the same. 
I don't think we are going to break anything - I'm not aware of any client relying on these access API urls in our DDIs.

**Comment by landreev** [comment]
Also, indexing global ids does have some value - I was wrong on that. Searching for an individual global id, once you already know it, doesn't make sense. BUT searching for all the files in a specific name space is often useful.

**Comment by landreev** [comment]
The Terms of Use issue: #2911 
The issue is 2 years old, but there's a recent comment from a specific partner, indicating that this is causing a problem for them.

**Comment by pameyer** [comment]
@landreev For JSON and DDI, I'd thought the recommendation for machine-readable formats was to use `doi:` instead of `http:`.  But if I'm wrong, we should probably be using `https` instead of `http`.

**Comment by adam3smith** [comment]
The guidance on this is confusing, but I'd essentially never use doi: anymore. So either the format allows something like `<identifier identifierType="DOI">10.3886/ICPSR03674.V2</identifier>` (from Datacite XML [here](https://data.datacite.org/application/vnd.datacite.datacite+xml/10.3886/icpsr03674.v2)) where DOI is specified in an attribute or you use the full resolved DOI in the form `https://doi.org/<doi>` as DataCite [does in their JSON-LD](https://data.datacite.org/application/vnd.schemaorg.ld+json/10.3886/icpsr03674.v2): 
`  "@id": "https://doi.org/10.3886/icpsr03674.v2",`

**Comment by landreev** [comment]
Hmm. We've been using http://dx.doi.org/10.7910/DVN/XXX (same for the handles equivalents) everywhere. 
what's the importance of https there? - they are just links to the global resolver... The actual registered urls that these resolve to are https urls of our dataset pages.

**Comment by adam3smith** [comment]
Detailed reason including tl:dr; at the top here: https://www.crossref.org/blog/linking-dois-using-https-the-background-to-our-new-guidelines/

The dx. isn't really problematic (and will keep working indefinitely) but it's unnecessary and discouraged by the DOI fundation, CrossRef, and DataCite, so you should take that out as well.

**Comment by landreev** [comment]
@adam3smith and everybody:
For the global id of the dataset we've been/we'll continue using 
```
<IDNo agency="DOI">doi:10.7910/DVN/XXXXX</IDNo>
```
in the DDI.
So we are leaving it to the user/software client to construct a resolvable url out of this identifier. 
I've been talking about the URLs we supply in the individual file sections  further down:
```
<fileDscr ID="fYYYYY" URI="https://dataverse.harvard.edu/api/access/datafile/YYYYY">
```
this is the direct access API url; and YYYYY is the internal database id of the file. (the "ID" attribute is the XML id that references the section; it's an opaque string - we don't need to worry about using a meaningful identifier there)
So I've been proposing replacing it with 
```
<fileDscr ID="fYYYYY" URI="http://dx.doi.org/10.7910/DVN/ZZZZZ">
```
where 10.7910/DVN/ZZZZZ is the registered DOI for the file...
So I'm not even suggesting we mess with how we are encoding any identifiers; we'd be replacing one URL with another...

**Comment by landreev** [comment]
... And I'm not opposed to still be adding the "doi:10.7910/DVN/ZZZZZ" version of the id somewhere else in that "fileDscr" section... Not sure where - the DDI does not seem to have  dedicated field for the identifier of an individual file... Could be a note: 
```
<note type="global id, file">doi:10.7910/DVN/ZZZZZ</note>
```

**Comment by pameyer** [comment]
@landreev - 

```
<fileDscr ID="fYYYYY" URI="http://dx.doi.org/10.7910/DVN/ZZZZZ">
```

un-confused me; thanks. My suggestion about using `doi:$foo` wouldn't make sense in this context.

**Comment by landreev** [comment]
OK, I spent more time than I was planning on that performance investigation today... 
Comparing the dev. and this branches, I can confirm that all the performance degradation in uploading and saving files does come from the registration calls. It sounds like doing it asynchronously would be something to explore... 

However, it appears that the EditDatafiles/upload files page is very seriously broken, already in the dev. branch, independently of the individual global ids for files. 

One standard test we used to run when optimizing that page was uploading and saving 1000 very small file. Not a very realistic test, but useful to detect serious performance penalties per single file. It used to work very fast. It no longer works at all. 

Something got broken in the rendering of the files list. You still have the 1000 files upload fairly quickly. Then the page starts rendering the file list, adding the files 1 by 1 and fully re-rendering the entire list every time... renders one file, erases the list, adds the second file, renders 2 files, erases... By the 300th or so file the page is unusable. 

So basically it's something you're not going to notice when uploading 1 or 3 files at a time. But it makes it impossible to upload large numbers of files at the same time. 

I don't know when it got broken. Some work was done on the page last summer; and some just last month. 

I don't think fixing this belongs in this ticket (totally unrelated to the file identifiers). But I feel like it needs to be fixed.

**Comment by djbrooke** [comment]
Thanks @landreev. I saw that you created #4452 to cover the files list performance issue and we'll handle it there. We should address the registration-call related performance issues before moving this one to done.

**Comment by pdurbin** [comment]
@sekmiller had tweaked `BatchImportIT` in 02bf719 and asked me to take a look. I made some improvements to the `roundTripDdi` method in b8090f0.

**Comment by landreev** [comment]
(adding @scolapasta)
There is definitely a reproducible database update conflict, when some of the files in question are being ingested, in parallel with the global id assignment. 
Even more importantly, there's another issue that @sekmiller identified yesterday - when a user clicks "publish" immediately after uploading the files, there can be a scenario of 2 global id assignment jobs racing against one another. 

The only clean way of doing this seems to be to handle both the global id assignment and ingest as one "post upload" task, running them in sequence and using the same ingest lock for the duration of the combined task. 

Let's discuss, if necessary.

**Comment by landreev** [comment]
In the light of this, I would maybe like to hear this explained one more time - why we insist on having these ids assigned on upload? 
Still seems like a potentially sensible approach to leave it till publish.

**Comment by landreev** [comment]
I made a commit disabling the asynchronous job for persistent ids. 
@sekmiller do you need to check in something else into this branch, or is it ready? 

Also, reviewed the implications of the functionality in this branch for *harvested* files. We do NOT want to assign persistent identifiers for harvested DataFiles, ever. It appears that we are safe. Harvested files are never created with CreateDataFileCommand outside CreateDatasetCommand. And the latter always calls the former with the skipAssigningIdentifier=true, like this:
```
ctxt.engine().submit(new CreateDataFileCommand(dataFile, dsv, getRequest(), dataFileIdentifier, true));//skip registration
```

But maybe I'll add an explicit check for (!isHarvested()) to the command, to be safe.

**Comment by landreev** [comment]
@sekmiller so which test(s) is failing now?

**Comment by sekmiller** [comment]
On FilesIT test_AddFileBadUploadFormat was failing - it’s really testing that you shouldn’t be able to upload a file via the api in the when the upload method is rsync and in the absence of a script. After consulting with Phil I commented out the test because it really shouldn’t be run without the DCM running in the background.

so now all tests are passing.


On Mar 27, 2018, at 2:40 PM, landreev <notifications@github.com<mailto:notifications@github.com>> wrote:


@sekmiller<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sekmiller&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=d7bbsLIcERTG2Qe_gE_iKH1z80JxaTHhJpRXaS4mxwM&m=WPn-hZZ4hRiQBEZ2lDhumpo6d6A2uRsVZjFORYnfMjY&s=R1VJJLKzjVfGY5e1Y7jlxjOCyRjMxz0E1da22dPQBZc&e=> so which test(s) is failing now?

**Comment by landreev** [comment]
@kcondon
@scolapasta asked me to help review the parts of the PR that specifically deal with locks. From looking at the code, it all looks good. But, as part of QA, please double-check that the Dataset page properly refreshes itself after the dataset becomes unlocked; (either after ingests finish, or when the persistent id registration is complete). 
That auto-refreshing has always been tricky. So every time we touch the locks, there is at least a possibility for it to stop working properly.

**Comment by pdurbin** [comment]
At standup I mentioned that pull request #4350 had merge conflicts. I merged the latest from develop into that branch which resulted in the SQL script getting renamed. In 639562f I fixed the SQL scripts so that upgrade_v4.8.5_to_v4.8.6.sql only has the 4.8.6 updates and a new script has the updates for 4.9.

**Comment by kcondon** [comment]
-no file cards on file upload/save to dataset
-exception and inability to publish dataset:
```Caused by: org.apache.solr.client.solrj.impl.HttpSolrClient$RemoteSolrException: Error from server at http://localhost:8983/solr/collection1: ERROR: [doc=datafile_49] unknown field 'filePersistentId
```

**Comment by landreev** [comment]
The solr error looks like a missing schema update.
"no file cards on file upload" - this could also be a result of merging this branch with my fix for the editfiles fragment?

**Comment by landreev** [comment]
Retracting what I just said about the file card - I thought you were talking about the file entry on the dataset page. 
If it was a search card then yes, it must be the solr schema issue.

**Comment by landreev** [comment]
@scolapasta @sekmiller 

Regarding the latest problem Kevin is seeing/reporting, above - that the performance is back to super unacceptable, *but only when large numbers of files are uploaded on Create* (1000 small files - takes 26 min. to save in Create Dataset mode; but only some number of seconds when uploading and saving the same files in an existing dataset). 
It really looks like this is the result of calling CreateDataFileCommand from inside CreateDataset command. When files are added to existing datasets, the CreateDataFileCommand is called by the page (and not from inside UpdateDatasetCommand). It looks like calling a command from inside another command results in the EJB trying to do everything as one big transaction; and transaction overhead gets worse (exponentially?) with the number of nested objects. So when uploading 100 files on Create, it is still acceptably fast (some fraction of a second per file); at 1000 files, it appears that it is taking 2 seconds per each CreateDataFileCommand. 

I'm guessing the solution is to move the CreateDataFile command out of CreateDatasetCommand. Save the dataset first, without files, then call CreateDataFile command for every file separately - like we do when files are added to existing datasets. 


Alternatively, maybe we could just junk the CreateDataFile command (it's new in this branch); now that we don't want to create/register file ids on create.

**Comment by sekmiller** [comment]
I refactored the createDatasetCommand so that it doesn't call the CreateDataFileCommand in a nested fashion.  When I tried the "1000 files" test creating the Dataset took less than a minute and publishing it took less than 10, but the Dataset Page was locked throughout the publish process and came back automatically when the publish was completed.

**Comment by landreev** [comment]
Adding relevant info about the current state of the branch; 
this is from @sekmiller via slack on Fri.:

```
we've seen a database connection error when the number of files on a dataset is well over 500. The logs show errors like: Severe:   RAR5031:System Exception
java.lang.NullPointerException
    at com.sun.enterprise.resource.ConnectorXAResource.getResourceHandle(ConnectorXAResource.java:246)
    at com.sun.enterprise.resource.ConnectorXAResource.end(ConnectorXAResource.java:159)
    at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.delistResource(JavaEETransactionManagerSimplified.java:527)
If you wait long enough the dataset will eventually publish and all of the files will be registered, however, the page is never locked to the user but at the end of the process the dataset gets it lock and it doesn't go away.
I put a hack into the Datapage init such that if a dataset is locked for PIDRegistration but there's no draft version, we try to remove the lock.
Is there a better way to approach this?
```
To me this looks very wrong. And I don't feel this really is about the right way of handling an exception - why are we getting that exception in the first place? Specifically, a database connection error? Are we somehow trying to register these file DOIs for the 500+ files *at the same time* - instead of doing int sequentially in the background? 
Steve also said that
```
They at least get identifiers in the database I didn't check that they are all actually registered at the EZID admin console, but a spot check shows that all the ones I have looked for there have been registered.
```
- so at least it's *looking* like that asynchronous job finishes doing what it needs to do, just never removes the locks properly... Still, this database exception, and the mess it leaves behind looks scary to me. And I don't think we should leave it like this, just because 500+ files is an "edge case".

**Comment by landreev** [comment]
I also added to specific change requests in the review. 
One for removing an Async attribute that doesn't seem to serve any purpose; and another for the part in the FinalizeDatasetPublicationCommand where we appear to be trying to delete the same lock twice... 
No good reason to think either of these 2 things are causing trouble, but who knows .

**Comment by landreev** [comment]
A quick summary, of the latest developments:
- the identifiers for files are not immediately assigned on upload; for as long as the file is in draft, it has no global id. so the db. ids are used in urls pointing to the file landing pages, etc. 
- the identifiers are assigned when the version is published. 
- if there are 10 or fewer files (the limit is configurable, but 10 is the default), the identifiers are assigned as part of the single publishing task. Meaning, you click publish, you (eventually) get the success message, all the files should have global identifiers. (if the global ids cannot be assigned, publish should fail). 
- if there are more than 10, it goes into the async mode. The version becomes locked, with a message clearly explaining what's going on. While locked, the version should still appear as an unpublished draft. Once the global ids are assigned in the background, the version should get automatically unlocked and appear as published. 
- So both the sync (10 or fewer files) and async modes should be tested separately. 
- Publishing datasets with large numbers of files should definitely be tested carefully. The expected behavior is that it should work with any number of files; with the async. part simply taking longer (the rule of thumb is that it costs .5 sec. per file to register the global id). But the last issue we ran into was that with more than 500 files or so, we were running into a weird concurrency issue where there was a conflict between the first (the synchronous) Publish command and the asynchronous part (the FinalizePublish command) that was messing up the locking and/or unlocking and sometimes preventing the dataset from getting published. It appears to have been fixed and everything should be working as advertised now.

**Comment by kcondon** [comment]
Tested the above checklist by leonid and all works as described. Also nearly completed test checklist: https://docs.google.com/document/d/1VOlDJFRy7zoS4LVDerH5TgMYsr3DKkm9ZRNcqajitxY/edit?usp=sharing
all works for ezid but handles and datacite do not register ids. handle does not create them, datacite now creates but does not register and is slow.
